kimchi/src/circuits/polynomials/foreign_field_add/witness.rs

```diff
@@ -127,74 +160,134 @@ pub fn create<F: PrimeField>(
 
     let foreign_modulus = ForeignElement::from_biguint(modulus);
 
-    // Create multi-range-check witness for first left input
     let mut left = ForeignElement::from_biguint(inputs[LO].clone());
-    range_check::witness::extend(&mut witness, left.clone());
-    let mut add_values: Vec<(F, F, F, F)> = vec![];
+    // Create multi-range-check witness for first left input
+    if range_checks {
+        range_check::witness::extend(&mut witness, left.clone());
+    }
+    let mut values: Vec<FFAddValues<F>> = vec![];
     for i in 0..num {
         let right = ForeignElement::from_biguint(inputs[i + 1].clone());
-        let (output, sign, overflow, carry_lo, carry_mi) =
-            compute_subadd_values(&left, &right, opcodes[i], &foreign_modulus);
-        // Create multi-range-check witness for right_input (left_input was done in previous iteration) and output
+        let add_values = compute_ffadd_values(&left, &right, opcodes[i], &foreign_modulus);
         // We only obtain the 3 lower limbs of right because the range check takes only 264 bits now
         let right_3_limb = ForeignElement::new([right[LO], right[MI], right[HI]]);
-        range_check::witness::extend(&mut witness, right_3_limb);
-        range_check::witness::extend(&mut witness, output.clone());
+        // Create multi-range-check witness for right_input (left_input was done in previous iteration) and output
+        if range_checks {
+            range_check::witness::extend(&mut witness, right_3_limb);
+            range_check::witness::extend(&mut witness, add_values.output.clone());
+        }
 
-        add_values.append(&mut vec![(sign, overflow, carry_lo, carry_mi)]);
-        left = output; // output
+        values.push(add_values.clone());
+        left = add_values.output.clone(); // output
     }
 
     // Compute values for final bound check, needs a 4 limb right input
     let right = ForeignElement::<F, 4>::from_biguint(BigUint::from(TWO_TO_LIMB).pow(3));
 
-    let (bound, sign, overflow, bound_carry_lo, bound_carry_mi) =
-        compute_subadd_values(&left, &right, FFOps::Add, &foreign_modulus);
+    let bound_values = compute_ffadd_values(&left, &right, FFOps::Add, &foreign_modulus);
     // Make sure they have the right value
-    assert_eq!(sign, F::one());
-    assert_eq!(overflow, F::one());
+    assert_eq!(bound_values.sign, F::one());
+    assert_eq!(bound_values.ovf, F::one());
```

Comment:
Would be cool to assert the other possible carries that we think are zero actually are.